### PR TITLE
Speed up environment

### LINKED ISSUE #577: TTP: some environments are much slower than the tf version
For example:
ac_simple_navigation.gin
ppo_bullet_humanoid.gin

### PR DESCRIPTION
**Pull request by emailweixu**
1. convert to cuda after stack timesteps from environments
2. support flatten for ParallelTorchEnvironment
3. always use numpy array before ParallelEnvironment

The time for each iteration of ppo_bullet_humanoid.gin is reduced from 14.33 sec to 8.98 sec.
Still 20% slower than tf version, which is 7.47.

### CODE REVIEW

File: alf/environments/alf_gym_wrapper.py [hunk 1]
```diff
@@ -145,12 +155,11 @@ def _obtain_zero_info(self):
...
         self._gym_env.reset()
-        action = nest.map_structure(lambda spec: spec.zeros(),
+        action = nest.map_structure(lambda spec: spec.numpy_zeros(),
                                     self._action_spec)
-        _, _, _, info = self._gym_env.step(self._convert_action(action))
+        _, _, _, info = self._gym_env.step(action)
         self._gym_env.reset()
-        return nest.map_structure(lambda i: torch.as_tensor(np.zeros_like(i)),
-                                  info)
+        return _as_array(info)
```
Comment (hnyu) [inline alf/environments/alf_gym_wrapper.py]: This is no longer zero info. It might affect the summary.

File: alf/tensor_specs.py [hunk 2]
```diff
@@ -69,6 +69,19 @@ def from_tensor(cls, tensor, from_dim=0):
         assert isinstance(tensor, torch.Tensor)
         return TensorSpec(tensor.shape[from_dim:], tensor.dtype)
 
+    @classmethod
+    def from_array(cls, array, from_dim=0):
+        """Create TensorSpec from numpy array.
+
+        Args:
+            tensor (np.ndarray|np.number): array from which the spec is extracted
```
Comment (hnyu) [inline alf/tensor_specs.py]: tensor -> array

File: alf/tensor_specs.py [hunk 3]
```diff
@@ -69,6 +69,19 @@ def from_tensor(cls, tensor, from_dim=0):
...
+            from_dim (int): use array.shape[from_dim:] as shape
```
Comment (hnyu) [inline alf/tensor_specs.py]: \``array.shape[from_dim:]\``

File: alf/tensor_specs.py [hunk 4]
```diff
@@ -144,6 +157,36 @@ def zeros(self, outer_dims=None):
...
+    def numpy_constant(self, value, outer_dims=None):
+        """Create a constant np.ndarray from the spec.
+
+        Args:
+            value : a scalar
+            outer_dims (tuple[int]): an optional list of integers specifying outer
+                dimensions to add to the spec shape before sampling.
+
+        Returns:
+            tensor (torch.Tensor): a tensor of ``self._dtype``.
```
Comment (hnyu) [inline alf/tensor_specs.py]: np.ndarray: an array of ``self._dtype``

File: alf/tensor_specs.py [hunk 5]
```diff
@@ -144,6 +157,36 @@ def zeros(self, outer_dims=None):
...
+
+    def numpy_zeros(self, outer_dims=None):
+        """Create a zero numpy.ndarray from the spec.
+
+        Args:
+            outer_dims (tuple[int]): an optional list of integers specifying outer
+                dimensions to add to the spec shape before sampling.
+
+        Returns:
+            tensor (torch.Tensor): a tensor of ``self._dtype``.
```
Comment (hnyu) [inline alf/tensor_specs.py]: np.ndarray: an array of ``self._dtype``

File: alf/trainers/policy_trainer.py [hunk 6]
```diff
@@ -192,6 +192,7 @@ def progress():
 
     def train(self):
         """Perform training."""
+        #os.system("taskset -p -c %s %d" % (','.join([str(i) for i in range(20, 32)]), os.getpid()))
```
Comment (hnyu) [inline alf/trainers/policy_trainer.py]: remove

**Review by hnyu — CHANGES_REQUESTED** [review]
Do you think it's necessary to write a short paragraph in "ALF knowledge base" about how we handle environment I/O (numpy, tensor, cuda/cpu, stacking) on the doc website?

File: alf/environments/alf_environment.py [hunk 7]
```diff
@@ -25,7 +25,7 @@
 
 
 @six.add_metaclass(abc.ABCMeta)
-class TorchEnvironment(object):
+class AlfEnvironment(object):
     """Abstract base class for Torch RL environments.
```
Comment (runjerry) [inline alf/environments/alf_environment.py:29]: Torch -> ALF ?

File: alf/environments/alf_environment.py [hunk 8]
```diff
@@ -25,7 +25,7 @@
...
```
Comment (runjerry) [inline alf/environments/alf_environment.py:30]: Torch -> ALF?

File: alf/environments/alf_gym_wrapper.py [hunk 9]
```diff
@@ -218,9 +218,19 @@ def _to_tensor_observation(self, observation):
         Returns:
             A (nested) tensors of observation
         """
```
Comment (runjerry) [inline alf/environments/alf_gym_wrapper.py:221]: Update comments accordingly.

File: alf/environments/alf_wrappers.py [hunk 10]
```diff
@@ -23,29 +23,27 @@
...
-from alf.utils.tensor_utils import to_tensor
 
 
-class TorchEnvironmentBaseWrapper(torch_environment.TorchEnvironment):
-    """TorchEnvironment wrapper forwards calls to the given environment."""
+class AlfEnvironmentBaseWrapper(AlfEnvironment):
+    """AlfEnvironment wrapper forwards calls to the given environment."""
 
     def __init__(self, env):
         """Create a torch environment base wrapper.
```
Comment (runjerry) [inline alf/environments/alf_wrappers.py]: torch -> ALF

File: alf/environments/alf_wrappers.py [hunk 11]
```diff
@@ -93,14 +94,14 @@ def wrapped_env(self):
 
 # Used in ALF
 @gin.configurable
-class TimeLimit(TorchEnvironmentBaseWrapper):
+class TimeLimit(AlfEnvironmentBaseWrapper):
     """End episodes after specified number of steps."""
 
     def __init__(self, env, duration):
         """Create a TimeLimit torch environment.
```
Comment (runjerry) [inline alf/environments/alf_wrappers.py]: torch -> ALF

File: alf/environments/alf_wrappers.py [hunk 12]
```diff
@@ -159,14 +160,14 @@ def _channel_transpose(tensor):
 
 
 @gin.configurable
-class ImageChannelFirst(TorchEnvironmentBaseWrapper):
+class ImageChannelFirst(AlfEnvironmentBaseWrapper):
     """Make images in observations channel_first. """
 
     def __init__(self, env):
         """Create a TimeLimit torch environment.
```
Comment (runjerry) [inline alf/environments/alf_wrappers.py]: torch -> ALF

File: alf/environments/suite_gym.py [hunk 13]
```diff
@@ -90,8 +90,8 @@ def wrap_env(gym_env,
     transpose, 'image_channel_first' is set as True by default. There are two options
     provided in ALF to handle this transpose:
         1. Applying the gym_wrappers.ImageChannelFirst after all gym_env_wrappers
-            and before the TorchGymWrapper.
-        2. Applying the torch_wrappers.ImageChannelFirst after all torch_gym_wrappers.
+            and before the AlfGymWrapper.
+        2. Applying the alf_wrappers.ImageChannelFirst after all alf_gym_wrappers.
     The first option is used in current function.
```
Comment (runjerry) [inline alf/environments/suite_gym.py]: The second option becomes meaningless now given that numpy array is used before and after AlfGymWrapper, so maybe only keep the following:

... 'image_channel_first' is set as True by default. gym_wrappers.ImageChannelFirst is applied after all gym_env_wrappers and before the AlfGymWrapper.

**Review by runjerry — COMMENTED** [review]
This is very neat way of keeping everything in numpy and on cpu until parallel env!

File: alf/tensor_specs.py (hunk 2 above)
Comment (emailweixu) [reply alf/tensor_specs.py]: Fixed

File: alf/environments/alf_gym_wrapper.py (hunk 1 above)
Comment (emailweixu) [reply alf/environments/alf_gym_wrapper.py]: Fixed

File: alf/tensor_specs.py (hunk 4 above)
Comment (emailweixu) [reply alf/tensor_specs.py]: fixed

File: alf/tensor_specs.py (hunk 5 above)
Comment (emailweixu) [reply alf/tensor_specs.py]: fixed

File: alf/environments/alf_gym_wrapper.py (hunk 9 above)
Comment (emailweixu) [reply alf/environments/alf_gym_wrapper.py:221]: Updated

File: alf/environments/suite_gym.py (hunk 13 above)
Comment (emailweixu) [reply alf/environments/suite_gym.py]: updated

File: docs/notes/knowledge_base.rst [hunk 14]
```diff
@@ -191,6 +192,62 @@ Step type      Discount | Value used          | Value          Note
...
+
+    gym_spec = gym.spec(environment_name)
+    gym_env = gym_spec.make()
+
+2. Apply a series of gym wrappers. One of the most often used gym wrapper is
+`ImageChannelFirst <../api/alf.environments.html#alf.environments.gym_wrappers.ImageChannelFirst>`_,
+which converts image with channel-last format to channel-first format. ALF
+uses channel-first format for its convolution layers.
+
+3. Wrap the gym environment as a on-batched ``AlfEnvironment`` using
```
Comment (hnyu) [inline docs/notes/knowledge_base.rst]: non-batched

File: docs/notes/knowledge_base.rst [hunk 15]
```diff
@@ -191,6 +192,62 @@ Step type      Discount | Value used          | Value          Note
...
+`AlfGymWrapper <../api/alf.environments.html#alf.environments.alf_gym_wrappers.AlfGymWrapper>`_.
+All of its input/output are ``numpy.ndarray``.
```
Comment (hnyu) [inline docs/notes/knowledge_base.rst]: inputs/outputs

File: docs/notes/knowledge_base.rst [hunk 16]
```diff
@@ -191,6 +192,62 @@ Step type      Discount | Value used          | Value          Note
...
+
+4. Apply a series of `ALF environment wrappers <../api/alf.environments.html#alf.environments.alf_wrappers>`_.
+All of its input/output are ``numpy.ndarray``.
+
+5. Wrap the non-batched ALF environmnet with `ProcessEnvironment <../api/alf.environments.html#alf.environments.process_environment.ProcessEnvironment>`_.
+It provides an interface using CPU torch.Tensor and interacts with the underline
+``AlfEnvironment`` using ``numpy.ndarray``.
+
+6. Use `ParallelEnvironment <../api/alf.environments.html#alf.environments.parallel_environment.ParallelEnvironment>`_
```
Comment (hnyu) [inline docs/notes/knowledge_base.rst:232]: Need to mention the conversion between tensor and array in this class?

Comment (emailweixu) [reply docs/notes/knowledge_base.rst:232]: This is done in ProcessEnvironment, which is mentioned in step 5.

**Review by hnyu — APPROVED** [review]

**Pull request merged**